### DISCUSSION TITLE
'NestedTensor' object has no attribute 'size'

### DISCUSSION THREAD

**Discussion by Aliweka2020** [rwightman/efficientdet-pytorch#255] [Q&A]
I am trying to use efficientnet bacbone with DETR but get this error

  File "C:\Users\Ali\AppData\Local\Continuum\anaconda3\lib\site-packages\timm\models\layers\padding.py", line 32, in pad_same
    ih, iw = x.size()[-2:]
AttributeError: 'NestedTensor' object has no attribute 'size'

I searched for NestedTensor to unpack to get the dimension but useless.

Could anyone help?!

**Comment by arekmula** [comment]
I have a similar problem. This function requires a single tensor, but somehow I pass a batch of tensors to this. In my case list of tensors, so I have 
```
    ih, iw = x.size()[-2:]
AttributeError: 'list' object has no attribute 'size'
```

**Comment by arekmula** [comment]
@Aliweka2020 So, I fixed this problem on my side. 

I use `train_one_epoch` function from [engine.py](https://github.com/pytorch/vision/blob/master/references/detection/engine.py). I needed to adapt those lines:
```
images = list(image.to(device) for image in images)
targets = [{k: v.to(device) for k, v in t.items()} for t in targets]

loss_dict = model(images, targets)
```
to this:
```
        images = torch.stack(images)
        boxes = [target["boxes"].float() for target in targets]
        labels = [target["labels"].float() for target in targets]
        target = {
            "bbox": boxes,
            "cls": labels
        }

        loss_dict = model(images, target)
```
because as I mentioned, the efficientdet requires tensor (Batch x Channels x Height x Width) not list or NestedTensor.

There are more possible fixes to this problem.
- You can write your own collate_fn function that would do this if you are not using `train_one_epoch`. 
- You can write a wrapper around model() that would transform the NestedTensor to Tensor and then forward the proper call.

**Reply by Aliweka2020** [reply]
Thanks  for your solution
I have tried your snippet of code and works for me, but I still stucked in training phase
I got this error after inserting effcientnet backbone
RuntimeError: Given groups=1, weight of size [256, 2048, 1, 1], expected input[1, 320, 24, 32] to have 2048 channels, but got 320 channels instead

after applying self.conv = nn.Conv2d(320, 2048, 1), I got this error
 TypeError: conv2d() received an invalid combination of arguments - got (list, Parameter, Parameter, tuple, tuple, tuple, int), but expected one of:
 * (Tensor input, Tensor weight, Tensor bias, tuple of ints stride, tuple of ints padding, tuple of ints dilation, int groups)
      didn't match because some of the arguments have invalid types: (list, Parameter, Parameter, tuple, tuple, tuple, int)

**Reply by rwightman** [reply]
Sorry, I have no idea what you are doing, the model factory and class takes care of handling config and connecting efficientnet backbones and has been tested for the models present in the config here.

**Reply by Aliweka2020** [reply]
![image](https://user-images.githubusercontent.com/6129944/142398941-f95ca96d-51a4-4a70-bbac-db88747b1d48.png)

**Reply by Aliweka2020** [reply]
This error means something to you, if you could help?

**Comment by rwightman** [comment]
I'm going to move this to discussions since it isn't a bug.

The workound @arekmula provided if you're using a training setup that uses lists or tuples is correct, the 'bench' wrappers for training here expect tensors. Another approach would be to write an alternate Train/Predict bench wrapper (https://github.com/rwightman/efficientdet-pytorch/blob/master/effdet/bench.py#L79)  that accept sequences. ie DetBenchTrainSeq / DetBenchPredictSeq ... and modify the factory create fn to accept 'train_seq' and 'predict_seq' as the bench type. I'd accept a PR for that.

For nested tensor, I haven't spent much time with it, but since it's out of PyTorch tree it's not a high priority for me to support but likely can be supported with a similar approach as above (specific bench wrappers).
